### ISSUE TITLE
base64 encode/decode of opaque binary data

### ISSUE THREAD

**Issue by tis-rpage** [mikefarah/yq#1112]
Please add support for base64 encoding/decoding of opaque binary data as documented in the YAML draft standard:
https://yaml.org/type/binary.html

yq already supports these conversion encodings:
Format | Decode (from string) | Encode (to string)
Yaml | from_yaml | to_yaml(i)/@yaml
JSON | from_json | to_json(i)/@json
Properties | | to_props/@props
CSV | | to_csv/@csv
TSV | | to_tsv/@tsv
XML | from_xml | to_xml(i)/@xml

Recomendation:
Base64 | from_base64 | to_base64(i)/@base64

**Comment by mikefarah** [comment]
Just thinking how this would work, I can imagine encoding doing something like:

```bash
yq '.image_data = load_as_base64("myimage.png")
```

Which would load the image file and encode it into the yaml as base64 string into the `image_data` node.

For decoding - it gets a little stranger as it doesn't really make sense (as far as I can tell) to have decoded binary data in an expression - that is, you can't really pipe it into another operator and make sense of it.

Not sure how that would work. Maybe its an output format option like:

```
yq '.image_data' -o=decode_base64 > image.png
```

thoughts? What are you trying to do with the binary data?

**Comment by tis-rpage** [comment]
kubernetes configmap and secrets are base64 encoded, so decoding the b64 data will yield the raw values which may be assignable to other text nodes.

I do the following constantly with jq, because yq doesn't have b64 encode/decode:
`kubectl get secret credential -o json | jq -r '.data[] | @base64d'`
so I'd want to be able to do this:
`kubectl get secret credential -o json | yq e '.data[] | @base64d'`

Dealing with x.509 certificates is another common use case where PEM content needs to be base64 encoded, and if there's raw DER somewhere, it'd be nice to be able to generate the PEM from a RAW der without shelling to openssl.

So perhaps the following might become a thing with using hashicorp vault for storing secrets:
`yq '.x509.cert = pem_header(load_as_base64("mycert.der"))' -o json | vault kv write secret/x509/mycert.pem @/dev/stdin`
 where pem header would represent some external mechanism for generating the PEM header/footer blocks

**Comment by baurmatt** [comment]
Also interested in this functionality. Mostly interested to work with Kubernetes Secrets.

**Comment by mikefarah** [comment]
Cheers - I'll work on this in the next release.

**Comment by mikefarah** [comment]
Released in 4.21.1

**Comment by mbigras** [comment]
Thanks for implementing! This [@base64d](https://mikefarah.gitbook.io/yq/operators/encode-decode) Base64 decode feature helped me decode Kubernetes secrets like the following example illustrates.

```
$ kubectl get secret mysecret -o yaml
apiVersion: v1
data:
  PASSWORD: MWYyZDFlMmU2N2Rm
  USER_NAME: YWRtaW4=
kind: Secret
# ...
$ kubectl get secret mysecret -o yaml | yq '.data | map_values(. | @base64d)'
PASSWORD: 1f2d1e2e67df
USER_NAME: admin
```

This is more ergonomic than the copy–paste suggestion in the [Decode the Secret](https://kubernetes.io/docs/tasks/configmap-secret/managing-secret-using-kubectl/#decoding-secret) Kubernetes documentation and also avoids the gotcha where you paste your secret in your shell history.
